polar-core/src/resource_block.rs

```diff
@@ -60,128 +44,147 @@ fn term_source_range(term: &Term) -> Range<usize> {
     start..end
 }
 
-pub fn validate_parsed_declaration(
-    (name, term): (Symbol, Term),
-) -> Result<Production, LalrpopError<usize, Token, error::ParseError>> {
-    match (name.0.as_ref(), term.value()) {
-        ("roles", Value::List(_)) => Ok(Production::Roles(term)),
-        ("permissions", Value::List(_)) => Ok(Production::Permissions(term)),
-        ("relations", Value::Dictionary(_)) => Ok(Production::Relations(term)),
+#[derive(Debug)]
+enum ParsedDeclaration {
+    Roles(Term),       // List<String>
+    Permissions(Term), // List<String>
+    Relations(Term),   // Dict<Symbol, Symbol>
+}
+
+fn validate_parsed_declaration((name, term): (Term, Term)) -> PolarResult<ParsedDeclaration> {
+    match (name.value().as_symbol()?.0.as_ref(), term.value()) {
+        ("roles", Value::List(_)) => Ok(ParsedDeclaration::Roles(term)),
+        ("permissions", Value::List(_)) => Ok(ParsedDeclaration::Permissions(term)),
+        ("relations", Value::Dictionary(_)) => Ok(ParsedDeclaration::Relations(term)),
 
         ("roles", Value::Dictionary(_)) | ("permissions", Value::Dictionary(_)) => {
             let (loc, ranges) = (term.offset(), vec![term_source_range(&term)]);
-            let msg = format!("Expected '{}' declaration to be a list of strings; found a dictionary:\n", name);
-            Err(LalrpopError::User { error: ParseError::ResourceBlock { loc, msg, ranges } })
+            let msg = format!("Expected '{}' declaration to be a list of strings; found a dictionary:\n", name.to_polar());
+            Err(ValidationError::ResourceBlock { loc, msg, ranges }.into())
```

Comment:
What if we stored the offending term on the resource block error instead? And then we could move all the location-finding logic elsewhere?

Reply:
82dd969